src/functions.js

```diff
@@ -0,0 +1,220 @@
+export function getScore() {
+  num1 = 100;
+  num2 = 600;
+  function add() {
+    return name + " scored " + (num1 + num2);
+  }
+  return add();
+}
+
+export function addSquares(a, b) {
+  function square(x) {
+    return x * x;
+  }
+  return square(a) + square(b);
+}
+
+//Function inside can acess function outside's  arguments and variables
+//When calling outside function, x is preserved when inside is returned
+export function outside(x) {
+  function inside(y) {
+    return x / y;
+  }
+  return inside;
+}
+
+//Multiple-nested functions
+//B forms a closure including A (B can access A's arguments and variables)
+//C forms a closure including B (C can access both B and A's arguments and variables)
+export function A(x) {
+  function B(y) {
+    function C(z) {
+      return x + y + z;
+    }
+    return C;
+  }
+  return B;
+}
+
+//When there's naming conflicts, the inner-most scope takes the highest precedence
+export function namingConflict() {
+  let x = 5;
+  function inside(x) {
+    return x * 2;
+  }
+  return inside;
+}
+
+//Another closure example
+export let createPet = function (name) {
+  var sex;
+
+  return {
+    setName: function (newName) {
+      name = newName;
+    },
+
+    getName: function () {
+      return name;
+    },
+
+    getSex: function () {
+      return sex;
+    },
+
+    setSex: function (newSex) {
+      if (
+        typeof newSex === "string" &&
+        (newSex.toLowerCase() === "male" || newSex.toLowerCase() === "female")
+      ) {
+        sex = newSex;
+      }
+    },
+  };
+};
+
+//Arguments object
+export function useArgumentsObject(separator) {
+  // separator will only take the first param
+  let result = "";
+  let i;
+  for (i = 1; i < arguments.length; i++) {
+    result += arguments[i] + separator;
+  }
+  return result;
+}
+
+//Function parameters
+//With default params
+export function useDefaultParams(a, b = 1) {
+  return a * b;
+}
+
+//With rest params
+export function useRestParams(multiplier, ...theArgs) {
+  return theArgs.map((x) => multiplier * x);
+}
+
+//Arrow functions
+// Arrow functions does not have its own this, arguments, super, or new.target
+// and are always anonymus
+export function useArrowFunction(words) {
+  return words.map((s) => s.length);
+}
+
+export const functionObject = {
+  x: 2,
+  y: 3,
+  arrowFunction: () => {
+    return this.x + this.y;
+  },
+  standardFunction: function () {
+    return this.x + this.y;
+  },
+  arrowThisInheritance: function () {
+    const values = [5, 6];
+    return values.map((v) => v * this.x);
+  },
+  standardThisInheritance: function () {
+    const values = [5, 6];
+    return values.map(function (v) {
+      v * this.x; // this.x is not valid because `this` is not passed down in a standard function
+    });
+  },
+};
+
+export class FunctionClass {
+  constructor() {
+    this.x = 2;
+    this.y = 3;
+    // this.arrowFunction = () => {
+    //   return this.x + this.y;
+    // };
+  }
```

Comment:
Not sure if that's what we are trying to do here, when I create an arrow function like this, it can get access to `this.x` and `this.y`. It has the same access when the `arrowFunction` is created in the constructor.

Reply:
Ya, the two syntaxes accomplish the same thing.

```js
class FunctionClass {
  x = 2;
  y = 3;
  arrowFunction = () => {  return this.x + this.y;  };
}
```

Will yield the same result as 

```js
class FunctionClass {
  constructor() {
    this.x = 2;
    this.y = 3;
    this.arrowFunction = () => { return this.x + this.y };
  }
}
```

The first syntax, which does not use the `constructor` function, is a short-hand syntax. The first syntax works great for anything that is constant in a class. With the `FunctionClass` in this code, we could do away with the constructor because no arguments are being passed into the constructor.